From: Maximilian Weigand <mweigand@mweigand.net>
Date: Tue, 25 Jul 2023 21:42:13 +0200
Subject: [rockchip_ebc/panel-simple] Change mode clocks

Now that we control the dclk clock using the dclk_select module
parameter, we can use the display modes to define realistic refresh
rates that refer to a full display refresh, not to one frame.

Add display refresh rates of 20, 25, 30, and 40 Hz.

This should decrease the pressure on the compositors, as they will try
to flip buffers with that frequency.
---
 drivers/gpu/drm/panel/panel-simple.c | 120 +++++++++++++++++++++++++++--------
 1 file changed, 92 insertions(+), 28 deletions(-)

diff --git a/drivers/gpu/drm/panel/panel-simple.c b/drivers/gpu/drm/panel/panel-simple.c
index 0f6b2f3..eabd2f8 100644
--- a/drivers/gpu/drm/panel/panel-simple.c
+++ b/drivers/gpu/drm/panel/panel-simple.c
@@ -2096,33 +2096,11 @@ static const struct panel_desc edt_etmv570g2dhu = {
 };
 
 static const struct drm_display_mode eink_ed103tc2_mode[] ={
+	/* clock frequency = refresh rate / (htotal * vtotal) / 1000 */
+	// 20 Hz
 	{
-		/* this will lead to dclk == 250 Mhz */
-		/* .clock = 266693, */
-		/* request dclk == 200 MHz */
-		.clock = 200000,
-		.hdisplay = 1872,
-		/* hsync_start, for linux timing order, should be display + fp
-		 * according to the data sheet, display = ldl == 234 [dclk] = 1872 pixels, and fp == lel == 7 [dclk] == 56 pixels
-		 */
-		/* data sheet days sync == lsl == 18 sclk = 18 * 8 pixels = 144 pixels */
-		.hsync_start = 1872 + 56,
-		.hsync_end = 1872 + 56 + 144,
-		.htotal = 1872 + 56 + 144 + 136,
-		/* .hsync_start = 1872 + 184, */
-		/* .hsync_end = 1872 + 184 + 88, */
-		/* .htotal = 1872 + 184 + 88 + 64, */
-		/* .hskew = 136, */
-		.hskew = 64,
-		.vdisplay = 1404,
-		.vsync_start = 1404 + 12,
-		.vsync_end = 1404 + 12 + 1,
-		.vtotal = 1404 + 12 + 1 + 4,
-		.flags = DRM_MODE_FLAG_PHSYNC | DRM_MODE_FLAG_NVSYNC |
-			 DRM_MODE_FLAG_HSKEW | DRM_MODE_FLAG_CLKDIV2,
-	},
-	{
-		.clock = 220000,
+		.clock = 62751,
+		/* .clock = 200000, */
 		.hdisplay = 1872,
 		.hsync_start = 1872 + 56,
 		.hsync_end = 1872 + 56 + 144,
@@ -2135,8 +2113,9 @@ static const struct drm_display_mode eink_ed103tc2_mode[] ={
 		.flags = DRM_MODE_FLAG_PHSYNC | DRM_MODE_FLAG_NVSYNC |
 			 DRM_MODE_FLAG_HSKEW | DRM_MODE_FLAG_CLKDIV2,
 	},
+	// 25 Hz
 	{
-		.clock = 266693,
+		.clock = 78439,
 		.hdisplay = 1872,
 		.hsync_start = 1872 + 56,
 		.hsync_end = 1872 + 56 + 144,
@@ -2149,11 +2128,96 @@ static const struct drm_display_mode eink_ed103tc2_mode[] ={
 		.flags = DRM_MODE_FLAG_PHSYNC | DRM_MODE_FLAG_NVSYNC |
 			 DRM_MODE_FLAG_HSKEW | DRM_MODE_FLAG_CLKDIV2,
 	},
+	// 30 Hz
+	{
+		.clock = 94127,
+		.hdisplay = 1872,
+		.hsync_start = 1872 + 56,
+		.hsync_end = 1872 + 56 + 144,
+		.htotal = 1872 + 56 + 144 + 136,
+		.hskew = 64,
+		.vdisplay = 1404,
+		.vsync_start = 1404 + 12,
+		.vsync_end = 1404 + 12 + 1,
+		.vtotal = 1404 + 12 + 1 + 4,
+		.flags = DRM_MODE_FLAG_PHSYNC | DRM_MODE_FLAG_NVSYNC |
+			 DRM_MODE_FLAG_HSKEW | DRM_MODE_FLAG_CLKDIV2,
+	},
+	// 40 Hz
+	{
+		.clock = 125502,
+		.hdisplay = 1872,
+		.hsync_start = 1872 + 56,
+		.hsync_end = 1872 + 56 + 144,
+		.htotal = 1872 + 56 + 144 + 136,
+		.hskew = 64,
+		.vdisplay = 1404,
+		.vsync_start = 1404 + 12,
+		.vsync_end = 1404 + 12 + 1,
+		.vtotal = 1404 + 12 + 1 + 4,
+		.flags = DRM_MODE_FLAG_PHSYNC | DRM_MODE_FLAG_NVSYNC |
+			 DRM_MODE_FLAG_HSKEW | DRM_MODE_FLAG_CLKDIV2,
+	},
+
+/* 	/1* modes that actually take the .clock seriously *1/ */
+/* 	{ */
+/* 		/1* this will lead to dclk == 250 Mhz *1/ */
+/* 		/1* .clock = 266693, *1/ */
+/* 		/1* request dclk == 200 MHz *1/ */
+/* 		.clock = 200000, */
+/* 		.hdisplay = 1872, */
+/* 		/1* hsync_start, for linux timing order, should be display + fp */
+/* 		 * according to the data sheet, display = ldl == 234 [dclk] = 1872 pixels, and fp == lel == 7 [dclk] == 56 pixels */
+/* 		 *1/ */
+/* 		/1* data sheet days sync == lsl == 18 sclk = 18 * 8 pixels = 144 pixels *1/ */
+/* 		.hsync_start = 1872 + 56, */
+/* 		.hsync_end = 1872 + 56 + 144, */
+/* 		.htotal = 1872 + 56 + 144 + 136, */
+/* 		/1* .hsync_start = 1872 + 184, *1/ */
+/* 		/1* .hsync_end = 1872 + 184 + 88, *1/ */
+/* 		/1* .htotal = 1872 + 184 + 88 + 64, *1/ */
+/* 		/1* .hskew = 136, *1/ */
+/* 		.hskew = 64, */
+/* 		.vdisplay = 1404, */
+/* 		.vsync_start = 1404 + 12, */
+/* 		.vsync_end = 1404 + 12 + 1, */
+/* 		.vtotal = 1404 + 12 + 1 + 4, */
+/* 		.flags = DRM_MODE_FLAG_PHSYNC | DRM_MODE_FLAG_NVSYNC | */
+/* 			 DRM_MODE_FLAG_HSKEW | DRM_MODE_FLAG_CLKDIV2, */
+/* 	}, */
+/* 	{ */
+/* 		.clock = 220000, */
+/* 		.hdisplay = 1872, */
+/* 		.hsync_start = 1872 + 56, */
+/* 		.hsync_end = 1872 + 56 + 144, */
+/* 		.htotal = 1872 + 56 + 144 + 136, */
+/* 		.hskew = 64, */
+/* 		.vdisplay = 1404, */
+/* 		.vsync_start = 1404 + 12, */
+/* 		.vsync_end = 1404 + 12 + 1, */
+/* 		.vtotal = 1404 + 12 + 1 + 4, */
+/* 		.flags = DRM_MODE_FLAG_PHSYNC | DRM_MODE_FLAG_NVSYNC | */
+/* 			 DRM_MODE_FLAG_HSKEW | DRM_MODE_FLAG_CLKDIV2, */
+/* 	}, */
+/* 	{ */
+/* 		.clock = 266693, */
+/* 		.hdisplay = 1872, */
+/* 		.hsync_start = 1872 + 56, */
+/* 		.hsync_end = 1872 + 56 + 144, */
+/* 		.htotal = 1872 + 56 + 144 + 136, */
+/* 		.hskew = 64, */
+/* 		.vdisplay = 1404, */
+/* 		.vsync_start = 1404 + 12, */
+/* 		.vsync_end = 1404 + 12 + 1, */
+/* 		.vtotal = 1404 + 12 + 1 + 4, */
+/* 		.flags = DRM_MODE_FLAG_PHSYNC | DRM_MODE_FLAG_NVSYNC | */
+/* 			 DRM_MODE_FLAG_HSKEW | DRM_MODE_FLAG_CLKDIV2, */
+/* 	}, */
 };
 
 static const struct panel_desc eink_ed103tc2 = {
 	.modes = eink_ed103tc2_mode,
-	.num_modes = 3,
+	.num_modes = 4,
 	.bpc = 4,
 	.size = {
 		.width = 210,
